pyanaconda/modules/users/users_interface.py

```diff
@@ -130,3 +120,100 @@ def IsRootAccountLocked(self) -> Bool:
+        # over DBUS
+        user_dicts = []
+
+        for user_struct in self.implementation.users:
+            user_dicts.append(get_structure(user_struct))
+        return user_dicts
+
+    @emits_properties_changed
+    def SetUsers(self, users: List[Structure]):
+        """Set a list of users, each corresponding to a single user.
+
+        :param users: a list of user describing DBUS structures
+        """
+        user_data_list = []
+        for user_struct in users:
+            user_data = self.implementation.create_user_data()
+            apply_structure(user_struct, user_data)
+            user_data_list.append(user_data)
+        self.implementation.set_users(user_data_list)
+
+    @property
+    def Groups(self) -> List[Structure]:
+        """List of groups, each describing a single group.
+
+        :return: a list of group describing DBUS Structures
+        """
+        # internally we hold the data about groups as a list of structures,
+        # which we need to turn into a list of dicts before returning it
+        # over DBUS
+        group_dicts = []
+
+        for group_struct in self.implementation.groups:
+            group_dicts.append(get_structure(group_struct))
+        return group_dicts
+
+    @emits_properties_changed
+    def SetGroups(self, groups: List[Structure]):
+        """Set a list of groups, each corresponding to a single group.
+
+        :param groups: a list of group describing DBUS structures
+        """
+        group_data_list = []
+        for group_struct in groups:
+            group_data = self.implementation.create_group_data()
+            apply_structure(group_struct, group_data)
+            group_data_list.append(group_data)
+        self.implementation.set_groups(group_data_list)
+
+    @property
+    def SshKeys(self) -> List[Structure]:
+        """List of SSH keys, each describing a single SSH key.
+
+        :return: a list of SSH key describing DBUS Structures
+        """
+        # internally we hold the data about SSH keys as a list of structures,
+        # which we need to turn into a list of dicts before returning it
+        # over DBUS
+        ssh_key_dicts = []
+
+        for ssh_key_struct in self.implementation.ssh_keys:
+            ssh_key_dicts.append(get_structure(ssh_key_struct))
+        return ssh_key_dicts
+
+    @emits_properties_changed
+    def SetSshKeys(self, ssh_keys: List[Structure]):
+        """Set a list of DBUS structures, each corresponding to a single SSH key.
+
+        :param ssh_keys: a list of SSH key describing DBUS structures
+        """
+        ssh_key_data_list = []
+        for ssh_key_struct in ssh_keys:
+            ssh_key_data = self.implementation.create_ssh_key_data()
+            apply_structure(ssh_key_struct, ssh_key_data)
+            ssh_key_data_list.append(ssh_key_data)
+        self.implementation.set_ssh_keys(ssh_key_data_list)
+
+    @property
+    def AdminUserExists(self) -> Bool:
```

Comment:
DBus properties might be expected to send a signal when their value is changing. Maybe I would use a method to avoid these expectations.

Reply:
Yeah, lets do that, at least for now. I've looked into what would be needed to properly trigger the changed signal for this aggregate property and it would be quite a lot of additional code to check every case when admin user can be added and removed. And each time you would have to remember what the previous state was and then compare with the state after the mutating change happened.

It could likely be done via a decorator, but still - rather keep it simple. :)